symengine/polynomial.h

```diff
@@ -8,9 +8,253 @@
 #include <symengine/expression.h>
 #include <symengine/monomials.h>
 
+// Calculates bit length of number, used in UIntDict*= only
+template <typename T>
+unsigned int bit_length(T t)
+{
+    unsigned int count = 0;
+    while (t > 0) {
+        count++;
+        t = t >> 1;
+    }
+    return count;
+}
+
 namespace SymEngine
 {
 
+class UIntDict
```

Comment:
Yes,  I was also thinking  the same lines. How should I approach this? Should I make a common template for dictionary wrappers, and inherit from them and add in special methods which are class specific?

 Also, can you remind me why `int` is used in `UnivariateExprPoly` and not `unsigned int`? 

Plus I feel we I should make an unordered map implementation of these classes. What do you think?